mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/binarystorage/BinaryStorageManager.java

```diff
@@ -0,0 +1,102 @@
+import java.util.Map;
+
+import org.hibernate.ogm.datastore.mongodb.logging.impl.Log;
+import org.hibernate.ogm.datastore.mongodb.logging.impl.LoggerFactory;
+import org.hibernate.ogm.datastore.mongodb.options.BinaryStorageType;
+import org.hibernate.ogm.datastore.mongodb.options.impl.BinaryStorageOption;
+import org.hibernate.ogm.datastore.mongodb.utils.TableEntityTypeMappingInfo;
+import org.hibernate.ogm.model.key.spi.EntityKey;
+import org.hibernate.ogm.model.spi.Tuple;
+import org.hibernate.ogm.options.spi.OptionsContext;
+import org.hibernate.ogm.options.spi.OptionsService;
+
+import com.mongodb.client.MongoDatabase;
+import org.bson.Document;
+
+/**
+ * @author Sergey Chernolyas &amp;sergey_chernolyas@gmail.com&amp;
+ */
+public class BinaryStorageManager {
+	private static final Log log = LoggerFactory.make( MethodHandles.lookup() );
+
+	private static final Map<BinaryStorageType,BinaryStorageDelegator> BINARY_STORAGE_DELEGATOR_MAP;
+	private static final BinaryStorageDelegator NULL_DELEGATOR = new NullDelegator();
+
+	static {
+		EnumMap<BinaryStorageType, BinaryStorageDelegator> map = new EnumMap<>( BinaryStorageType.class );
+		map.put( BinaryStorageType.GRID_FS, new GridFSDelegator() );
+		BINARY_STORAGE_DELEGATOR_MAP = Collections.unmodifiableMap( map );
+	}
+
+	public static void storeContentToBinaryStorage(MongoDatabase mongoDatabase, Document currentDocument,
+											EntityKey entityKey, OptionsService optionService, Tuple tuple) {
+		Class entityClass = TableEntityTypeMappingInfo.getEntityClass( entityKey.getTable() );
+		if ( currentDocument == null ) {
+			return;
+		}
+		for ( String fieldName : currentDocument.keySet() ) {
+			if ( fieldName.equals( "$set" ) ) {
+				// it is part of request. it is not document
+				Document queryFields = (Document) currentDocument.get( fieldName );
+				for ( String queryField : queryFields.keySet() ) {
+					storeContentFromFieldToBinaryStorage( mongoDatabase, queryFields, entityClass, queryField, optionService, tuple );
+				}
+			}
+			else {
+				// it is not document
+				storeContentFromFieldToBinaryStorage( mongoDatabase, currentDocument, entityClass, fieldName, optionService, tuple );
+			}
+		}
+	}
+
+	private static void storeContentFromFieldToBinaryStorage(MongoDatabase mongoDatabase,  Document currentDocument, Class entityClass, String fieldName,
+									OptionsService optionService,Tuple tuple) {
+		OptionsContext optionsContext = getPropertyOptions( optionService, entityClass, fieldName );
+		BinaryStorageType binaryStorageType = optionsContext.getUnique( BinaryStorageOption.class );
+		BinaryStorageDelegator binaryStorageDelegator = BINARY_STORAGE_DELEGATOR_MAP.getOrDefault( binaryStorageType,NULL_DELEGATOR  );
+		binaryStorageDelegator.storeContentToBinaryStorage( mongoDatabase, optionsContext, currentDocument, fieldName,  tuple );
+	}
+
+	public static void removeFromGridFsByEntity(MongoDatabase mongoDatabase, OptionsService optionService, Document deletedDocument,
```

Comment:
There shouldn't be any mention of GridFS in this class because it's the manager of possible several storage type